src/loader-new.js

```diff
@@ -0,0 +1,342 @@
+   */
+  addLogo_(logo) {
+    this.svgRoot_.appendChild(logo);
+  }
+
+  /**
+   * Adds the background shim under the loader for cases where loader is on
+   * top of an image.
+   * @private
+   */
+  maybeAddBackgroundShim_() {
+    if (!this.hasImagePlaceholder_()) {
+      return;
+    }
+
+    // Not Implemented
+
+    return;
+  }
+
+  /**
+   * Add a gray default placeholder if there isn't a placeholder already and
+   * other special cases.
+   * @private
+   */
+  maybeAddDefaultPlaceholder_() {
+    // NOTE(aghassemi): I do believe we need to exclude amp-list here, but
+    // let's see how experimentation goes. Maybe a better idea is to have a
+    // white list, any component that does not fully load a new background may
+    // look bad if a gray placeholder shows up and goes away quickly. This
+    // default placeholder is good for image, video, etc.. but amp-list which
+    // usually just loads text is debatable. amp-iframe is also a candidate
+    // to exclude, often it does load a video or maps but it may not load
+    // text on transparent background in certain cases.
+    if (!hasPlaceholder(this.element_)) {
+      this.loadingContainer_.classList.add('i-amphtml-default-placeholder');
+    }
+  }
+
+  /**
+   * Whether the element is an Ad.
+   * @private
+   * @return {boolean}
+   */
+  isAd_() {
+    // Not Implemented
+    return false;
+  }
+
+  /**
+   * Whether the element is small.
+   * Small elements get a different loader with does not have a logo and is
+   * just a spinner.
+   * @private
+   * @return {boolean}
+   */
+  isSmall_() {
+    const box = this.element_.getLayoutBox();
+    return !isTiny(this.element_) && (box.width <= 100 || box.height <= 100);
+  }
+
+  /**
+   * Some components such as Instagram require larger spinner due to
+   * branding guidelines.
+   * @private
+   * @return {boolean}
+   */
+  requiresLargeSpinner_() {
+    // Not Implemented
+    return false;
+  }
+
+  /**
+   * Whether element has an existing image placeholder.
+   * @private
+   * @return {boolean}
+   */
+  hasImagePlaceholder_() {
+    // Not Implemented
+    return false;
+  }
+}
+
+/**
+ * Whether the new loader experiment is enabled.
+ * @param {!Window} win
+ * @return {boolean}
+ */
+export function isNewLoaderExperimentEnabled(win) {
+  return isExperimentOn(win, 'new-loaders');
+}
+
+/**
+ * Whether the element is eligible for loaders based on new loader heuristics
+ * This is is called by `isLoadingEnabled_` in `custom-element.js`
+ * and is additional heuristics to what `isLoadingEnabled_` checks.
+ * @param {!AmpElement} element
+ * @return {boolean}
+ */
+export function isLoaderIneligible(element) {
+  const result = isTiny(element) || isImageWithPlaceholder(element);
+  return result;
+}
+
+/**
+ * Very small layout are not eligible for new loaders.
+ * @param {!AmpElement} element
+ * @return {boolean}
+ */
+function isTiny(element) {
+  const box = element.getLayoutBox();
```

Comment:
This has the potential to cause a recalc, so should be put in a measure phase. The caller (`prepareLoading_`) happens in a mutate.

Reply:
getLayoutBox() returns the cached layout rect.

Reply:
In layers, it can cause a measure.

Reply:
There are lots of places that call `getLayoutBox` out of measure assuming it returns cached value.

Reply:
@jridgewell There is a `getLayoutWidth_` in custom_element that is also exposed through `getLayoutWidth()` which seems to be cached and used in similar situations ( https://github.com/ampproject/amphtml/blob/master/src/custom-element.js#L1636 ), should I add a `height` version and use that instead?

Reply:
Yes, that would be perfect. We could also discuss exposing cached versions in layers, but that'll probably not be compatible with the current resources measurements.

Reply:
> In layers, it can cause a measure.

Higher risk of layout thrashing?